### PR TITLE
Upgrading transformers version to 5.5.4

### PR DESCRIPTION
**Pull request by filyp**
# What does this PR do?
- It upgrades transformers to 5.5.4, to support newer models and features. Especially running MoE models much more efficiently (added in transformers 5).
- Also bumps other dependencies to versions required by `transformers==5.5.4`. (I extensively used this setup in my unlearning experiments, without any issues.)
- Adds a regression test for `prediction_step`, to make future upgrades easier and catch issues.
- Adds a dockerfile and a readme link to a prebuilt docker image with full environment (makes it easier to start using open-unlearning, especially on cloud GPUs that require specifying an image with the environment; helps reproducibility). 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Have you gone through the contributions [guide](../docs/contributing.md)?
- [ ] Are your changes documented? Read documentation guidelines [here](../README.md#-further-documentation).

## Tests
I used the newly added `prediction_step` regression test to verify validity; I ran `make quality`. I also run `python src/train.py --config-name=unlearn.yaml experiment=unlearn/tofu/default eval=tofu_simple question_key=paraphrased_question eval.tofu.batch_size=16 trainer.args.report_to=wandb trainer=NPO task_name=transformers5.5.4` to regression test with runs in https://github.com/locuslab/open-unlearning/pull/175. Like before, the reported loss changes (due to a different scaling convention in new version), but the actual unlearning trajectory stays almost the same.

image

One thing worth flagging: in the newer transformers versions (not only this one), installing flash-attn is more messy. Simple pip install will trigger a long build process (which also can fail if local CUDA version mismatches); there are prebuild wheels, but only in 3rd party repos (I documented it in readme). I think the cleanest solution would be to actually remove flash-attn, given that with typical unlearning datasets it actually *slows training down*, as discussed in https://github.com/locuslab/open-unlearning/issues/190

### CODE REVIEW

File: tests/prediction_step_regression.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    # On older transformers (no num_items_in_batch normalization) this was 2.7516,
+    # Due to difference in num of tokens used for loss normalization.
+    expected_loss = 2.4076595306396484
+
+    assert abs(loss.item() - expected_loss) < 1e-3, (loss.item(), expected_loss)
+    assert tuple(logits.shape) == expected_logits_shape
+    assert tuple(labels.shape) == expected_labels_shape
+    head = logits[0, 0, :8].tolist()
+    for got, exp in zip(head, expected_logits_head):
+        assert abs(got - exp) < 1e-3, (got, exp)
```
[the remark this answers is not in the document]
Comment (filyp) [reply tests/prediction_step_regression.py:88]: It actually *does* use tolerance based assersts (lines 89-94).

File: src/trainer/unlearn/base.py [hunk 2]
```diff
@@ -78,10 +67,10 @@ def _prepare_deepspeed(self, model):
     def prediction_step(
         self,
         model: nn.Module,
-        inputs: dict[str, Union[torch.Tensor, Any]],
+        inputs: dict[str, torch.Tensor | Any],
```
[the remark this answers is not in the document]
Comment (filyp) [reply src/trainer/unlearn/base.py:70]: Copied from original transformers code, must stay the same.

File: src/trainer/unlearn/base.py [hunk 3]
```diff
@@ -148,9 +135,18 @@ def prediction_step(
...
                 if has_labels or loss_without_labels:
                     with self.compute_loss_context_manager():
-                        ### Call compute_loss of super class since overridden compute_loss is not applicable to eval_dataset.
+                        num_items_in_batch = self._get_num_items_in_batch(
+                            [inputs], self.args.device
+                        )
+                        # !!!!!!! Call compute_loss of super class since overridden compute_loss is not applicable to eval_dataset.
+                        # loss, outputs = self.compute_loss(
+                        #     model, inputs, return_outputs=True, num_items_in_batch=num_items_in_batch
+                        # )
```
[the remark this answers is not in the document]
Comment (filyp) [reply src/trainer/unlearn/base.py:144]: It must be clear which part was swapped.

File: tests/prediction_step_regression.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    torch.manual_seed(SEED)
+
+    tokenizer = AutoTokenizer.from_pretrained(MODEL_NAME)
+    if tokenizer.pad_token_id is None:
+        tokenizer.pad_token = tokenizer.eos_token
+
+    model = AutoModelForCausalLM.from_pretrained(
+        MODEL_NAME, torch_dtype=torch.float32, attn_implementation="sdpa"
+    )
+    model.eval()
```
[the remark this answers is not in the document]
Comment (filyp) [reply tests/prediction_step_regression.py]: Meant as manual test, not CI.

**Comment by filyp** [comment]
What do you think @molereddy @Dornavineeth ?

On a more general note, I have a grant to work on LLM unlearning, and I could spend some portion of it helping to develop this open-unlearning repo.
I'm working in a fork where I have many things I could merge, like evaluation under relearning attacks, hyperparameter tuning with hydra, more methods, models and datasets. I also plan to build some leaderboard, and keep developing the models and datasets.

I know you have limited time to review contributions, so let me know what would be the most useful for me to do.

File: tests/prediction_step_regression.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (filyp) [reply tests/prediction_step_regression.py]: Now using non-gated 0.5B Qwen

File: README.md [hunk 5]
```diff
@@ -115,7 +115,9 @@ We provide several variants for each of the components in the unlearning pipelin
 conda create -n unlearning python=3.11
 conda activate unlearning
 pip install ".[lm-eval]"
-pip install --no-build-isolation flash-attn==2.6.3
+pip install --no-build-isolation flash-attn==2.8.3
+# Or to avoid building flash-attn:
+pip install "https://github.com/mjun0812/flash-attention-prebuild-wheels/releases/download/v0.7.16/flash_attn-2.8.3+cu128torch2.9-cp311-cp311-linux_x86_64.whl"
```
[the remark this answers is not in the document]
Comment (filyp) [reply README.md:120]: Ideally, we could remove flash-attn as proposed in https://github.com/locuslab/open-unlearning/issues/190

File: README.md [hunk 6]
```diff
@@ -125,6 +127,8 @@ python setup_data.py --eval # saves/eval now contains evaluation results of the
 # python setup_data.py --help
 ```
 
+We also provide a [Docker image](https://hub.docker.com/r/filyp/open-unlearning), with this environment already installed.
```
[the remark this answers is not in the document]
Comment (filyp) [reply README.md:130]: I'd rather not specify it, in case someone forgets to update readme. The hub page already provides that recommended tag to pull.
